### PR TITLE
Add adaptive federated optimization

### LINKED ISSUE #281: Adaptive Federated Optimization algorithms implementation
Implement inside OpenFL `FedYogi`, `FedAdagrad`, `FedAdam` aggregation algorithms from https://arxiv.org/pdf/2003.00295.pdf

### PR DESCRIPTION
**Pull request by alexey-khorkin**
This PR adds an implementation of adaptive optimizers (based on numpy `ndarray`) and tests for it.
Paper: https://arxiv.org/abs/2003.00295

TODO:
- [x] Add `Adagrad` optimizer
- [x] Add `Adam` optimizer
- [x] Add `Yogi` optimizer
- [x] Add more info like links to the original papers and docstrings inside `__init__` methods
- [x] Add another function to verify the correctness of optimizers inside tests
- [x] Add adaptive optimization aggregated functions
- [x] Add docs and edit one of existing tutorial

### CODE REVIEW

**Comment by itrushkin** [comment]
@alexey-khorkin Could you clarify the purpose of introducing numpy-based optimizers instead of using PyTorch-based ones?

**Comment by alexey-khorkin** [comment]
> @alexey-khorkin Could you clarify the purpose of introducing numpy-based optimizers instead of using PyTorch-based ones?

@itrushkin, This is necessary to get rid of the dependence of the PyTorch installation on the aggregator (director) side

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+"""Adagrad optimizer module."""
+
+import typing as tp
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: Please don't use such aliases

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class Adagrad(Optimizer):
+    """Adagrad optimizer implementation.
+
+    Original paper: http://jmlr.org/papers/v12/duchi11a.html
+    """
+
+    def __init__(self,
+                 params: tp.Dict[str, np.ndarray],
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: ~~tp.~~ `Dict.[str,np.ndarray]`

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                 learning_rate: float = 0.01,
+                 initial_accumulator_value: float = 0.1,
+                 epsilon: float = 1e-10,
+                 ) -> None:
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: Could you please to use such style for functions:
```python 
def __init__(
    self,
    param: str,
) -> None:
    self.param = param

```

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        super().__init__()
+
+        if learning_rate < 0:
+            raise ValueError(f'Invalid learning rate: {learning_rate}')
+        if initial_accumulator_value < 0:
+            raise ValueError(
+                f'Invalid initial_accumulator_value value: {initial_accumulator_value}')
+        if epsilon < 0:
+            raise ValueError(
+                f'Invalid epsilon value: {epsilon}')
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: It would be great if you add explanation what value is expected. 
Example:
```python
f'Invalid learning rate: {learning_rate}. Learning rate must be > 0.'
```
or
```python
f'Learning rate must be > 0, given: {learning_rate}'
```

Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: `learning_rate == 0`, `initial_accumulator_value == 0` and `epsilon == 0` are valid?

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        Implement Adagrad optimizer weights update rule.
+
+        Args:
+            gradients: Partial derivatives with respect to optimized parameters.
+        """
+        for grad_name in gradients:
+            grad = gradients[grad_name]
+
+            if grad_name not in self.grads_squared:
+                raise KeyError(f"Key {grad_name} doesn't exist in optimized parameters")
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py:82]: Looks like it should be on the top of `for` loop

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+            self.grads_squared[grad_name] = self.grads_squared[grad_name] + grad**2
+
+            # Make an update for a group of parameters
+            self.params[grad_name] = (self.params[grad_name] - self.learning_rate * grad
+                                      / (np.sqrt(self.grads_squared[grad_name]) + self.epsilon))
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: Maybe make sense to create function `update_param` or smth like this. It should increase readability and allow to remove comment.

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

**Comment by itrushkin** [comment]
> > @alexey-khorkin Could you clarify the purpose of introducing numpy-based optimizers instead of using PyTorch-based ones?
> 
> @itrushkin, This is necessary to get rid of the dependence of the PyTorch installation on the aggregator (director) side

Optimizer steps are usually performed on the collaborator side during a training task. Why do we need them on the aggregator?

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py (hunk 4 above)
Comment (alexey-khorkin) [reply openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: Learning rate == 0 is valid, epsilon ==0 is not, thanks!

**Comment by alexey-khorkin** [comment]
Docs render:
image

File: openfl/component/aggregation_functions/adaptive_aggregation.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
+# Copyright (C) 2021-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+"""Apartive aggregation module."""
```
Comment (itrushkin) [inline openfl/component/aggregation_functions/adaptive_aggregation.py]:
```suggestion
"""Adaptive aggregation module."""
```

File: openfl/component/aggregation_functions/adaptive_aggregation.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+        if tensor_name not in self.optimizer.params:
+            return self.default_agg_func(local_tensors,
+                                         db_iterator,
+                                         tensor_name,
+                                         fl_round,
+                                         tags)
+        else:
+            assert base_model_nparray is not None, (
+                f'There is no current global model in TensorDB for tensor name: {tensor_name}')
```
Comment (itrushkin) [inline openfl/component/aggregation_functions/adaptive_aggregation.py]: We usually avoid using `assert`s outside `tests` module. Please change this block to `if`-`raise` clause.

File: openfl/component/aggregation_functions/adaptive_aggregation.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                and 'delta' not in record['tags']
+            ):
+                base_model_nparray = record['nparray']
+
+        if tensor_name not in self.optimizer.params:
+            return self.default_agg_func(local_tensors,
+                                         db_iterator,
+                                         tensor_name,
+                                         fl_round,
+                                         tags)
```
Comment (itrushkin) [inline openfl/component/aggregation_functions/adaptive_aggregation.py]: This check can be moved to the top of the `call` function.

File: openfl/interface/interactive_api/experiment.py [hunk 10]
```diff
@@ -506,7 +506,7 @@ class ModelInterface:
     There is no support for several models / optimizers yet.
     """
 
-    def __init__(self, model, optimizer, framework_plugin) -> None:
+    def __init__(self, model, optimizer, framework_plugin: str) -> None:
```
Comment (itrushkin) [inline openfl/interface/interactive_api/experiment.py]: This change is not related to adaptive optimization, and only partially describes this method signature. I think this can be safely removed. Otherwise, typing should be applied to all constructor arguments.

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+from typing import Dict
+from typing import Optional
+
+import numpy as np
+
+from .base_optimizer import Optimizer
+
+
+class NumpyAdagrad(Optimizer):
```
Comment (itrushkin) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: NumPy is official name of the library. We should follow this style.

File: openfl/utilities/optimizers/numpy/adam_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+from typing import Dict
+from typing import Optional
+from typing import Tuple
+
+import numpy as np
+
+from .base_optimizer import Optimizer
+
+
+class NumpyAdam(Optimizer):
```
Comment (itrushkin) [inline openfl/utilities/optimizers/numpy/adam_optimizer.py]: NumPy is official name of the library. We should follow this style.

File: openfl/utilities/optimizers/numpy/adam_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    """Adagrad optimizer implementation.
```
Comment (itrushkin) [inline openfl/utilities/optimizers/numpy/adam_optimizer.py]:
```suggestion
    """Adam optimizer implementation.
```

File: openfl/utilities/optimizers/numpy/__init__.py [hunk 14]
```diff
@@ -0,0 +1,10 @@
+"""Numpy optimizers package."""
```
Comment (itrushkin) [inline openfl/utilities/optimizers/numpy/__init__.py:4]: Please add copyright lines to all files that have been added.
```
# Copyright (C) 2021-2022 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

**Review by itrushkin — CHANGES_REQUESTED** [review]

**Comment by dmitryagapov** [comment]
Maybe make sense to use this library https://github.com/ddbourgin/numpy-ml.
If we implement `AdaptiveAggregation` using this library we can simply add support for `RMSProp` and `SGD`.
I didn't find `Yogi` optimizer there, but we can create pull request to `numpy-ml`.

File: docs/overriding_agg_fn.rst [hunk 15]
```diff
@@ -32,8 +32,30 @@ Choose from the following predefined aggregation functions:
...
+
+``AdaptiveAggregation`` usage example:
+
+.. code-block:: python
+
+    from openfl.component.aggregation_functions import AdaptiveAggregation
+    from openfl.utilities.optimizers.numpy import NumpyAdagrad
+
+    # Creating aggregation function
+    agg_fn = AdaptiveAggregation(optimizer=NumpyAdagrad(model_interface=MI, learning_rate=0.4))
```
Comment (dmitryagapov) [inline docs/overriding_agg_fn.rst]: From `openfl.component.aggregation_function` should be imported aggregation function but not aggregation function "factory".
```python
from openlf.component.aggregation_functions import AdagradAdaptiveAggregation

agg_fn = AdagradAdaptiveAggregation(model_interace=MI, learning_rate=0.4)
```

File: docs/overriding_agg_fn.rst [hunk 16]
```diff
@@ -32,8 +32,30 @@ Choose from the following predefined aggregation functions:
...
+
+    See the `NumpyAdagrad <https://github.com/intel/openfl/blob/develop/openfl/utilities/optimizers/numpy/adagrad_optimizer.py>`_ and `AdaptiveAggregation <https://github.com/intel/openfl/blob/develop/openfl/component/aggregation_functions/adaptive_aggregation.py>`_ definitions for details.
+
+    `Adaptive federated optimization <https://arxiv.org/pdf/2003.00295.pdf>`_ original paper.
+
+``AdaptiveAggregation`` usage example:
+
+.. code-block:: python
+
+    from openfl.component.aggregation_functions import AdaptiveAggregation
```
Comment (dmitryagapov) [inline docs/overriding_agg_fn.rst]: Maybe it should be in `aggregation_functions.core`.

File: openfl/component/aggregation_functions/__init__.py [hunk 17]
```diff
@@ -3,9 +3,14 @@
...
 from .interface import AggregationFunction
 from .median import Median
 from .weighted_average import WeightedAverage
 
-__all__ = ['Median', 'WeightedAverage', 'GeometricMedian', 'AggregationFunction']
+__all__ = ['Median',
+           'WeightedAverage',
+           'GeometricMedian',
+           'AggregationFunction',
+           'AdaptiveAggregation']
```
Comment (dmitryagapov) [inline openfl/component/aggregation_functions/__init__.py]: Looks like it is not aggregation functions.

File: openfl/component/aggregation_functions/adaptive_aggregation.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+class AdaptiveAggregation(AggregationFunction):
+    """Adaptive Federated Aggregation funtcion.
+
+    According to https://arxiv.org/abs/2003.00295
+    """
+
+    def __init__(self, optimizer: Optimizer):
+        """Initialize."""
+        self.optimizer = optimizer
+        self.default_agg_func = WeightedAverage()
```
Comment (dmitryagapov) [inline openfl/component/aggregation_functions/adaptive_aggregation.py]: If we can't change it maybe it should be class attribute?

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+        assert model_interface is not None or params, (
+            'Should provide one of the params or model_interface')
+
+        if learning_rate < 0:
+            raise ValueError(
+                f'Invalid learning rate: {learning_rate}. Learning rate must be >= 0.')
+        if initial_accumulator_value < 0:
+            raise ValueError(
+                f'Invalid initial_accumulator_value value: {initial_accumulator_value}. \
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: Please don't use backslash.

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+            model_interface: Model interface instance to provide parameters.
+            learning_rate: Tuning parameter that determines
+                the step size at each iteration.
+            initial_accumulator_value: Initial value for squared gradients.
+            epsilon: Value for computational stability.
+        """
+        super().__init__()
+
+        assert model_interface is not None or params, (
+            'Should provide one of the params or model_interface')
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py]: Could you please replace it with `raise`

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+        Implement Adagrad optimizer weights update rule.
+
+        Args:
+            gradients: Partial derivatives with respect to optimized parameters.
+        """
+        for grad_name in gradients:
+            if grad_name not in self.grads_squared:
+                raise KeyError(f"Key {grad_name} doesn't exist in optimized parameters")
+
+            grad = gradients[grad_name]
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adagrad_optimizer.py:84]: `numpy-ml` having such code

```python
        # scale gradient to avoid explosion
        t = np.inf if clip_norm is None else clip_norm
        if norm(param_grad) > t:
            param_grad = param_grad * t / norm(param_grad)
```
does it have sense?

File: openfl/utilities/optimizers/numpy/adam_optimizer.py [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+
+            self._update_first_moment(grad_name, grad)
+            self._update_second_moment(grad_name, grad)
+
+            grads_first_moment_normalized = (self.grads_first_moment[grad_name]
+                                             / (1. - self.beta_1
+                                             ** (self.current_step[grad_name] + 1)))
+            grads_second_moment_normalized = (self.grads_second_moment[grad_name]
+                                              / (1. - self.beta_2
+                                              ** (self.current_step[grad_name] + 1)))
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adam_optimizer.py]: Maybe make sense create normalization function with short variables names because it looks unreadable.

File: openfl/utilities/optimizers/numpy/adam_optimizer.py [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+
+            # Make an update for a group of parameters
+            self.params[grad_name] = (self.params[grad_name]
+                                      - self.learning_rate
+                                      * grads_first_moment_normalized
+                                      / (np.sqrt(grads_second_moment_normalized) + self.epsilon))
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/adam_optimizer.py]:
```python 
self.params[grad_name] -= lr * momentum_1 / (np.sqrt(momentum_2) + eps)
```
or
```python
update = lr * momentum_1 / (np.sqrt(momentum_2) + eps)
self.params[grad_name] -= update 
```

File: openfl/utilities/optimizers/numpy/yogi_optimizer.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+    """Yogi optimizer implementation.
+
+    Original paper:
+    https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization
+    """
+
+    def __init__(
+        self,
+        params: Optional[Dict[str, np.ndarray]] = None,
+        *,
```
Comment (dmitryagapov) [inline openfl/utilities/optimizers/numpy/yogi_optimizer.py:21]: Why are you using `*` not after `self`?

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

File: openfl/component/aggregation_functions/__init__.py (hunk 17 above)
Comment (alexey-khorkin) [reply openfl/component/aggregation_functions/__init__.py]: @dmitryagapov `AdaptiveAggregation` is an aggregation function now, `AggregationFunction` already in develop branch 
https://github.com/intel/openfl/blob/e91ed7e32e85f38c692445d01b334c2357e0e322/openfl/component/aggregation_functions/__init__.py#L11

File: openfl/utilities/optimizers/numpy/yogi_optimizer.py (hunk 24 above)
Comment (alexey-khorkin) [reply openfl/utilities/optimizers/numpy/yogi_optimizer.py:21]: @dmitryagapov, because `params` variable can be passed by position, not only by keyword, it's for highlighting  that 
the 'params' has more importance and if both the `params` and the `model_interface` are passed, then the initialization will be through the `params`.

File: openfl/utilities/optimizers/numpy/adagrad_optimizer.py (hunk 21 above)
Comment (alexey-khorkin) [reply openfl/utilities/optimizers/numpy/adagrad_optimizer.py:84]: @dmitryagapov by default `clip_norm` is None - 
https://github.com/ddbourgin/numpy-ml/blob/b0359af5285fbf9699d64fd5ec059493228af03e/numpy_ml/neural_nets/optimizers/optimizers.py#L60
I can add this parameter, but I'm not sure if it's really needed.

File: openfl/utilities/optimizers/numpy/adam_optimizer.py (hunk 23 above)
Comment (alexey-khorkin) [reply openfl/utilities/optimizers/numpy/adam_optimizer.py]: @dmitryagapov, this change may produce error like this - https://techoverflow.net/2019/05/22/how-to-fix-numpy-typeerror-cannot-cast-ufunc-subtract-output-from-dtypefloat64-to-dtypeint64-with-casting-rule-same_kind/
if `self.params[grad_name].dtype == np.float(64)`. 
It can happen there is a model in int8.

File: openfl/component/aggregation_functions/adagrad_adaptive_aggregation.py [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+
+DEFAULT_AGG_FUNC = WeightedAverage()
+
+
+class AdagradAdaptiveAggregation(AdaptiveAggregation):
+    """Adagrad adaptive Federated Aggregation funtcion."""
+
+    def __init__(
+        self,
+        default_agg_func: AggregationFunction = DEFAULT_AGG_FUNC,
```
Comment (dmitryagapov) [inline openfl/component/aggregation_functions/adagrad_adaptive_aggregation.py]: Looks like it should be `agg_func`.

File: openfl/component/aggregation_functions/adagrad_adaptive_aggregation.py [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+        **kwargs
```
Comment (dmitryagapov) [inline openfl/component/aggregation_functions/adagrad_adaptive_aggregation.py]: I think that we should show to the user all params that could be used. Now user should go deep to the openFL code to understand which params could be `change`/`used`/`what is default value` for adaptive aggregation function. Of course amount of code will increate a little bit, but it will be more clear.

File: openfl/component/aggregation_functions/core/adaptive_aggregation.py [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+class AdaptiveAggregation(AggregationFunction):
+    """Adaptive Federated Aggregation funtcion.
+
+    According to https://arxiv.org/abs/2003.00295
+    """
+
+    def __init__(
+        self,
+        optimizer: Optimizer,
+        default_agg_func: AggregationFunction,
```
Comment (dmitryagapov) [inline openfl/component/aggregation_functions/core/adaptive_aggregation.py]: It looks like `agg_func`.

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

File: docs/overriding_agg_fn.rst [hunk 28]
```diff
@@ -32,12 +32,125 @@ Choose from the following predefined aggregation functions:
...
+                model_interface: Model interface instance to provide parameters.
+                learning_rate: Tuning parameter that determines
+                    the step size at each iteration.
+                param1: My own defined parameter.
+                param2: My own defined parameter.
+            """
+            super().__init__()
+            pass # Your code here!
+
+        def step(self, gradients: Dict[str, np.ndarray]) -> None:
```
Comment (igor-davidyuk) [inline docs/overriding_agg_fn.rst:116]: could you provide a scenario when `gradients` is a string?

File: docs/overriding_agg_fn.rst [hunk 29]
```diff
@@ -32,12 +32,125 @@ Choose from the following predefined aggregation functions:
...
+    from openfl.component.aggregation_functions import AdagradAdaptiveAggregation
+
+    TI = TaskInterface()
+    MI = ModelInterface(model=model,
+                        optimizer=optimizer,
+                        framework_plugin=framework_adapter)
+    ...
+
+    # Creating aggregation function
+    agg_fn = AdagradAdaptiveAggregation(model_interface=MI,
```
Comment (igor-davidyuk) [inline docs/overriding_agg_fn.rst:74]: Is it correct that adaptive aggregation and a custom aggregation function are mutually exclusive?
If we imagine a set of linear functions that merge multiple local updates to a global one (aggregation)
`global_model = L(local_models) = L (global(-1 round) + local updates) = global(-1 round) + L(local updates)`
It seems like the aggregation functions and optimizers may coexist

**Comment by igor-davidyuk** [comment]
> Maybe make sense to use this library https://github.com/ddbourgin/numpy-ml. If we implement `AdaptiveAggregation` using this library we can simply add support for `RMSProp` and `SGD`. I didn't find `Yogi` optimizer there, but we can create pull request to `numpy-ml`.

This question

File: docs/overriding_agg_fn.rst (hunk 28 above)
Comment (alexey-khorkin) [reply docs/overriding_agg_fn.rst:116]: @igor-davidyuk According to mypy [docs](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html):

```python3
# For mappings, we need the types of both keys and values
x: dict[str, float] = {"field": 2.0}  # Python 3.9+
x: Dict[str, float] = {"field": 2.0}
```

**Comment by alexey-khorkin** [comment]
> > Maybe make sense to use this library https://github.com/ddbourgin/numpy-ml. If we implement `AdaptiveAggregation` using this library we can simply add support for `RMSProp` and `SGD`. I didn't find `Yogi` optimizer there, but we can create pull request to `numpy-ml`.
> 
> This question

@igor-davidyuk , with @dmitryagapov  and @alexey-gruzdev, we have decided not to use this library for now.

**Review by dmitryagapov — APPROVED** [review]

**Review by itrushkin — APPROVED** [review]

**Review by igor-davidyuk — APPROVED** [review]

**Review by alexey-gruzdev — APPROVED** [review]
